File: tools/council.py
#!/usr/bin/env python3
"""A council of cheap cores, combined as a product of experts.

The single ridge probe reaches 69.4% on the independent EVAL set, and its
remaining errors are semantic rather than lexical -- cp -> same, du -> ls,
sysbox -> fsck. Those are cases where one mean-pooled vector genuinely cannot
separate two meanings, which argues for a second *view* of the request rather
than more examples of the first.

Hinton's products of experts (1999/2002) is the combination rule, not a mixture.
In a mixture one expert wins and the rest are discarded; in a product every
expert multiplies its opinion in, so a core that is confident sharpens the
result and a core that has no idea emits a near-flat distribution and
effectively abstains. That is the "cores that balance each other" behaviour,
and it falls out of summing log-probabilities rather than needing a gate.

Each core is given a genuinely different view, because two experts looking at
the same features disagree only by noise:

  semantic   mean-pooled embeddings, ridge. Knows that "duplicate" and "copy"
             are related. Blind to which exact word was used.
  lexical    multinomial naive Bayes over token counts. Knows exactly which
             words appeared and nothing about what they mean. Trained by
             counting, so there is no matrix to invert.
  character  hashed character 3-grams, same naive Bayes. Sees morphology --
             "duplicating" and "duplicate" share most of their trigrams even
             though they may be different tokens entirely.

Per-core weights are the temperature knob: a weight near zero makes a core
abstain regardless of how loud it is. They are swept here rather than assumed.
"""

import json
import sys
from collections import Counter
from pathlib import Path

import numpy as np

sys.path.insert(0, str(Path(__file__).parent))
import evaluate as ev  # noqa: E402
from reference import load  # noqa: E402

MODEL = "tools/out/smollm2-135m-q8.bin"
TOKENIZER = "tools/out/smollm2-tok.bin"
DATA = "tools/out/tools-dataset.json"
ev.HF_TOKENIZER = "tools/hf/tokenizer.json"


def log_softmax(M):
    M = M - M.max(axis=1, keepdims=True)
    return M - np.log(np.exp(M).sum(axis=1, keepdims=True))


# --- semantic core ------------------------------------------------------

def semantic(w, tr_ids, te_ids, ytr, C, lam=1.0):
    dim = w["embed"].shape[1]

    def pool(ids_list):
        out = np.zeros((len(ids_list), dim), dtype=np.float32)
        for i, ids in enumerate(ids_list):
            if ids:
                out[i] = w["embed"][ids].mean(axis=0)
        return out

    A, B = pool(tr_ids), pool(te_ids)
    mu = A.mean(axis=0)
    A, B = A - mu, B - mu
    Y = np.eye(C)[ytr]
    W = np.linalg.solve(A.T @ A + lam * np.eye(dim), A.T @ Y)
    return A @ W, B @ W


# --- counting cores -----------------------------------------------------

def naive_bayes(tr_counts, te_counts, ytr, C, V, alpha=0.2):
    """Multinomial naive Bayes. Training is counting; there is nothing to solve.

    Returns log-probabilities directly, which is what the product wants -- no
    calibration step needed, unlike the ridge scores.
    """
    counts = np.full((C, V), alpha, dtype=np.float64)
    for row, y in zip(tr_counts, ytr):
        for f, n in row.items():
            counts[y, f] += n
    logp = np.log(counts / counts.sum(axis=1, keepdims=True))

    prior = np.bincount(ytr, minlength=C).astype(np.float64)
    prior = np.log(prior / prior.sum())

    def score(rows):
        out = np.zeros((len(rows), C))
        for i, row in enumerate(rows):
            s = prior.copy()
            for f, n in row.items():
                s += n * logp[:, f]
            out[i] = s
        return out

    return score(tr_counts), score(te_counts)


def token_counts(ids_list, vocab):
    out = []
    for ids in ids_list:
        c = Counter()
        for t in ids:
            if t in vocab:
                c[vocab[t]] += 1
        out.append(dict(c))
    return out


def fnv1a(s):
    """Deterministic 32-bit hash.

    Python's `hash` is salted per process unless PYTHONHASHSEED is set, so
    using it here would make the character core's buckets -- and therefore its
    accuracy -- change between runs. It also has to be reimplementable in the
    kernel, and FNV-1a is a dozen lines anywhere.
    """
    h = 0x811C9DC5
    for b in s.encode("utf-8"):
        h = ((h ^ b) * 0x01000193) & 0xFFFFFFFF
    return h


def char_counts(texts, buckets=4096, n=3):
    out = []
    for t in texts:
        s = f"  {t.lower()}  "
        c = Counter()
        for i in range(len(s) - n + 1):
            c[fnv1a(s[i:i + n]) % buckets] += 1
        out.append(dict(c))
    return out


def accuracy(scores, y):
    return float((scores.argmax(1) == y).mean())


def main():
    cfg, w, _ = load(MODEL)
    tok = ev.Tok(TOKENIZER)
    data = json.loads(Path(DATA).read_text(encoding="utf-8"))
    train, test = data["train"], data["test"]
    names = ev.NAMES
    idx = {n: i for i, n in enumerate(names)}
    C = len(names)

    tr_text = [e["task"] for e in train]
    te_text = [e["task"] for e in test]
    ytr = np.array([idx[e["applet"]] for e in train])
    yte = np.array([idx[e["applet"]] for e in test])

    tr_ids = [ev.encode(tok, t) for t in tr_text]
    te_ids = [ev.encode(tok, t) for t in te_text]

    print(f"  {len(train)} train / {len(test)} test, {C} classes, "
          f"chance {100/C:.1f}%, resolution {100/len(test):.1f}%\n")

    cores = {}

    _, sem_te = semantic(w, tr_ids, te_ids, ytr, C)
    cores["semantic"] = log_softmax(sem_te)

    # Only tokens actually seen in training get a column; the rest can never
    # carry evidence and would only add smoothing mass.
    seen = sorted({t for ids in tr_ids for t in ids})
    vmap = {t: i for i, t in enumerate(seen)}
    _, lex_te = naive_bayes(token_counts(tr_ids, vmap), token_counts(te_ids, vmap),
                            ytr, C, len(seen))
    cores["lexical"] = log_softmax(lex_te)

    B = 4096
    _, chr_te = naive_bayes(char_counts(tr_text, B), char_counts(te_text, B),
                            ytr, C, B)
    cores["character"] = log_softmax(chr_te)

    print("  each core alone:")
    for name, s in cores.items():
        print(f"    {name:12} {accuracy(s, yte):6.1%}")

    # --- the product -----------------------------------------------------
    #
    # Weights are chosen by cross-validation on the *training* set and then
    # applied to EVAL exactly once. Sweeping a 216-point grid against EVAL and
    # reporting the maximum is not an estimate of anything: with 49 items it
    # selects noise, and it did -- the best combination that way dropped the
    # semantic core entirely, which no honest procedure supports.
    print("\n  products:")
    grid = [0.0, 0.25, 0.5, 0.75, 1.0, 1.5]
    folds = 5
    rng = np.random.default_rng(0)
    order = rng.permutation(len(ytr))

    fold_scores = {k: [] for k in cores}
    fold_y = []
    for f in range(folds):
        held = order[f::folds]
        keep = np.setdiff1d(order, held)
        sub_ids = [tr_ids[i] for i in keep]
        sub_y = ytr[keep]
        h_ids = [tr_ids[i] for i in held]

        _, s_sem = semantic(w, sub_ids, h_ids, sub_y, C)
        fold_scores["semantic"].append(log_softmax(s_sem))

        s_seen = sorted({t for ids in sub_ids for t in ids})
        s_map = {t: i for i, t in enumerate(s_seen)}
        _, s_lex = naive_bayes(token_counts(sub_ids, s_map),
                               token_counts(h_ids, s_map), sub_y, C, len(s_seen))
        fold_scores["lexical"].append(log_softmax(s_lex))

        _, s_chr = naive_bayes(char_counts([tr_text[i] for i in keep], B),
                               char_counts([tr_text[i] for i in held], B),
                               sub_y, C, B)
        fold_scores["character"].append(log_softmax(s_chr))
        fold_y.append(ytr[held])

    cv = {k: np.concatenate(v) for k, v in fold_scores.items()}
    cv_y = np.concatenate(fold_y)

    best = (None, -1.0)
    for a in grid:
        for b in grid:
            for c in grid:
                if a == b == c == 0:
                    continue
                s = a * cv["semantic"] + b * cv["lexical"] + c * cv["character"]
                acc = accuracy(s, cv_y)
                if acc > best[1]:
                    best = ((a, b, c), acc)
    a, b, c = best[0]
    print(f"    weights {best[0]} chosen by {folds}-fold CV on train "
          f"({best[1]:.1%} there)")

    chosen = a * cores["semantic"] + b * cores["lexical"] + c * cores["character"]
    print(f"    council on EVAL             {accuracy(chosen, yte):6.1%}")

    equal = sum(cores.values())
    print(f"    equal weights on EVAL       {accuracy(equal, yte):6.1%}")
    print(f"    semantic alone (baseline)   {accuracy(cores['semantic'], yte):6.1%}")

    # --- answer rules ----------------------------------------------------
    #
    # A product is not the only way to use three cores, and the kernel turned
    # up a case that made the distinction concrete: for "delete that file
    # please" the probe answered cat while both counting cores answered rm.
    # Under "the probe answers" that request routes wrongly even though two
    # cores out of three had it right. Majority vote is a different rule and
    # has to be measured rather than reasoned about.
    print("\n  answer rules:")
    picks = np.stack([cores[k].argmax(1) for k in ("semantic", "lexical", "character")])

    def majority(prefer_row):
        out = np.empty(len(yte), dtype=int)
        for i in range(len(yte)):
            col = picks[:, i]
            counts = Counter(col.tolist())
            top, n = counts.most_common(1)[0]
            # Ties go to the preferred core rather than to whichever class
            # happens to sort first.
            out[i] = top if n > 1 else col[prefer_row]
        return out

    for label, pred in (
        ("probe answers (current)", picks[0]),
        ("majority, ties to probe", majority(0)),
        ("majority, ties to lexical", majority(1)),
    ):
        print(f"    {label:28} {float((pred == yte).mean()):6.1%}")
    print(f"    {'product, equal weights':28} {accuracy(equal, yte):6.1%}")

    # The subset that matters: where the probe stands alone against the other
    # two. If the pair is usually right there, deferring to them is free
    # accuracy; if not, the current rule is correct as it stands.
    alone = (picks[0] != picks[1]) & (picks[1] == picks[2])
    if alone.any():
        probe_ok = float((picks[0][alone] == yte[alone]).mean())
        pair_ok = float((picks[1][alone] == yte[alone]).mean())
        print(f"\n  probe outvoted 2-1 on {int(alone.sum())} items:")
        print(f"    probe right there  {probe_ok:6.1%}")
        print(f"    pair right there   {pair_ok:6.1%}")

    # --- where the cores disagree ----------------------------------------
    # The interesting property is not only the score. If the cores agree on
    # most items, a cascade can answer those instantly and escalate only the
    # rest, which is the whole point of having more than one.
    picks = {k: s.argmax(1) for k, s in cores.items()}
    unanimous = np.ones(len(yte), dtype=bool)
    for k in picks:
        unanimous &= picks[k] == picks["semantic"]
    n_un = int(unanimous.sum())
    if n_un:
        print(f"\n  all three agree on {n_un}/{len(yte)} items, "
              f"{accuracy(np.eye(C)[picks['semantic'][unanimous]], yte[unanimous]):.1%} correct there")
    if n_un < len(yte):
        d = ~unanimous
        s = sum(cores.values())
        print(f"  they disagree on {int(d.sum())}, "
              f"{accuracy(s[d], yte[d]):.1%} correct after the product")


if __name__ == "__main__":
    main()

File: src/ai/council.rs
//! Several cheap cores, used to tell when the answer is probably wrong.
//!
//! The obvious thing to do with multiple cores is combine their opinions into
//! a better answer. That was measured and it does not work: on 108 independent
//! items the best single core scores 77.8% and an equal-weight product of
//! three scores 76.9%. So this does not do that.
//!
//! What the same measurement did show is that the cores' *agreement* predicts
//! correctness sharply. Where all three pick the same applet they are right
//! 90.3% of the time; where they split, 50%. That is a confidence signal, and
//! a considerably more useful thing to own than the point or two the product
//! was supposed to buy -- a router that knows when it is guessing can ask,
//! escalate, or refuse, and one that is silently 78% accurate cannot.
//!
//! So the semantic probe answers, being the strongest core alone, and the
//! other two corroborate. Their disagreement never changes the answer; it
//! changes what the system says about the answer.
//!
//! The two extra cores are multinomial naive Bayes, trained by counting. No
//! matrix to factorise, and each sees something the ridge probe cannot:
//!
//!   lexical    exact token identity. Knows "duplicate" was the word used, and
//!              nothing about what it means.
//!   character  hashed character trigrams over raw bytes. Sees morphology --
//!              "duplicating" and "duplicate" share most trigrams even when
//!              they tokenise differently -- and needs no tokenizer at all,
//!              which makes it independent of that whole layer.

use super::tensor::lnf;
use super::tokenizer::Tokenizer;
use alloc::vec;
use alloc::vec::Vec;

/// Buckets for the character core. Collisions are tolerable: a trigram sharing
/// a bucket with an unrelated one adds noise, not a wrong answer, and 4096
/// keeps the table at 21 classes * 4096 * 4 bytes.
pub const CHAR_BUCKETS: usize = 4096;

/// Laplace smoothing. Without it a feature unseen for some class gives that
/// class a log-probability of negative infinity and one unfamiliar word
/// eliminates it outright.
const ALPHA: f32 = 0.2;

/// Deterministic 32-bit hash.
///
/// FNV-1a rather than anything cleverer because it has to match the Python
/// side exactly, and because a salted hash would make the buckets -- and so
/// the accuracy -- differ between runs. That is not hypothetical: the first
/// measurement of this core used Python's `hash`, which is salted per process,
/// and reported four points that vanished when it was pinned.
pub fn fnv1a(bytes: &[u8]) -> u32 {
    let mut h: u32 = 0x811C_9DC5;
    for b in bytes {
        h ^= *b as u32;
        h = h.wrapping_mul(0x0100_0193);
    }
    h
}

/// Sparse multinomial naive Bayes.
pub struct Bayes {
    classes: usize,
    features: usize,
    /// `classes * features` log-probabilities.
    logp: Vec<f32>,
    log_prior: Vec<f32>,
}

impl Bayes {
    pub fn fit(
        docs: &[Vec<(u32, f32)>],
        labels: &[usize],
        classes: usize,
        features: usize,
    ) -> Option<Self> {
        if docs.is_empty() || classes == 0 || features == 0 || docs.len() != labels.len() {
            return None;
        }
        let mut counts = vec![ALPHA; classes * features];
        let mut per_class = vec![0.0f32; classes];

        for (doc, &y) in docs.iter().zip(labels.iter()) {
            if y >= classes {
                return None;
            }
            per_class[y] += 1.0;
            for &(f, n) in doc {
                let f = f as usize;
                if f < features {
                    counts[y * features + f] += n;
                }
            }
        }

        let mut logp = vec![0.0f32; classes * features];
        for c in 0..classes {
            let total: f32 = counts[c * features..(c + 1) * features].iter().sum();
            if total <= 0.0 {
                return None;
            }
            for f in 0..features {
                logp[c * features + f] = lnf(counts[c * features + f] / total);
            }
        }

        let docs_total: f32 = per_class.iter().sum();
        let log_prior = per_class
            .iter()
            .map(|n| lnf((n + ALPHA) / (docs_total + ALPHA * classes as f32)))
            .collect();

        Some(Self { classes, features, logp, log_prior })
    }

    pub fn scores(&self, doc: &[(u32, f32)]) -> Vec<f32> {
        let mut out = self.log_prior.clone();
        for &(f, n) in doc {
            let f = f as usize;
            if f >= self.features {
                continue;
            }
            for c in 0..self.classes {
                out[c] += n * self.logp[c * self.features + f];
            }
        }
        out
    }

    /// Argmax restricted to `allowed`.
    ///
    /// Permission is enforced by omission in every core, not only in the one
    /// that happens to answer. A corroborating core that voted for a mutating
    /// applet under read-only trust could otherwise carry a majority to
    /// something the operator forbade.
    pub fn predict_among(&self, doc: &[(u32, f32)], allowed: &[usize]) -> Option<usize> {
        let s = self.scores(doc);
        let mut best: Option<usize> = None;
        for &i in allowed {
            if i >= self.classes {
                continue;
            }
            if best.map(|b| s[i] > s[b]).unwrap_or(true) {
                best = Some(i);
            }
        }
        best
    }

    pub fn params(&self) -> usize {
        self.logp.len() + self.log_prior.len()
    }
}

/// Character trigrams of `text`, hashed into buckets.
///
/// Lowercased and padded, so the first and last characters of a request carry
/// the same kind of evidence as the middle ones.
pub fn char_features(text: &str) -> Vec<(u32, f32)> {
    let mut padded = Vec::with_capacity(text.len() + 4);
    padded.extend_from_slice(b"  ");
    for b in text.bytes() {
        padded.push(b.to_ascii_lowercase());
    }
    padded.extend_from_slice(b"  ");

    let mut counts: Vec<(u32, f32)> = Vec::new();
    for w in padded.windows(3) {
        let bucket = fnv1a(w) % CHAR_BUCKETS as u32;
        match counts.binary_search_by_key(&bucket, |(b, _)| *b) {
            Ok(i) => counts[i].1 += 1.0,
            Err(i) => counts.insert(i, (bucket, 1.0)),
        }
    }
    counts
}

/// The council: one probe that answers, two counters that corroborate.
pub struct Council {
    /// Token ids seen during training, sorted. Anything outside carries no
    /// evidence and would only dilute the smoothing mass.
    lex_vocab: Vec<u32>,
    lexical: Bayes,
    character: Bayes,
}

impl Council {
    /// `texts` and `labels` are the corpus; the semantic probe is fitted
    /// separately and passed to `vote`.
    pub fn fit(
        texts: &[&str],
        labels: &[usize],
        classes: usize,
        tok: &Tokenizer,
    ) -> Option<Self> {
        let ids: Vec<Vec<usize>> = texts.iter().map(|t| tok.encode(t, false, false)).collect();

        let mut lex_vocab: Vec<u32> = Vec::new();
        for row in &ids {
            for t in row {
                let v = *t as u32;
                if let Err(i) = lex_vocab.binary_search(&v) {
                    lex_vocab.insert(i, v);
                }
            }
        }
        if lex_vocab.is_empty() {
            return None;
        }

        let lex_docs: Vec<Vec<(u32, f32)>> = ids
            .iter()
            .map(|row| dense_counts(row, &lex_vocab))
            .collect();
        let lexical = Bayes::fit(&lex_docs, labels, classes, lex_vocab.len())?;

        let chr_docs: Vec<Vec<(u32, f32)>> =
            texts.iter().map(|t| char_features(t)).collect();
        let character = Bayes::fit(&chr_docs, labels, classes, CHAR_BUCKETS)?;

        Some(Self { lex_vocab, lexical, character })
    }

    pub fn params(&self) -> usize {
        self.lexical.params() + self.character.params()
    }

    /// What each corroborating core thinks, restricted to `allowed`.
    pub fn corroborate(
        &self,
        text: &str,
        tok: &Tokenizer,
        allowed: &[usize],
    ) -> Option<(usize, usize)> {
        let ids = tok.encode(text, false, false);
        let lex = self.lexical.predict_among(&dense_counts(&ids, &self.lex_vocab), allowed)?;
        let chr = self.character.predict_among(&char_features(text), allowed)?;
        Some((lex, chr))
    }
}

fn dense_counts(ids: &[usize], vocab: &[u32]) -> Vec<(u32, f32)> {
    let mut out: Vec<(u32, f32)> = Vec::new();
    for t in ids {
        let Ok(i) = vocab.binary_search(&(*t as u32)) else { continue };
        let i = i as u32;
        match out.binary_search_by_key(&i, |(f, _)| *f) {
            Ok(k) => out[k].1 += 1.0,
            Err(k) => out.insert(k, (i, 1.0)),
        }
    }
    out
}

/// What the system is willing to say about an answer.
pub struct Verdict {
    pub applet: usize,
    /// How many of the three cores chose it, including the probe itself.
    pub agreement: usize,
    /// What the corroborating cores said, when they disagreed.
    pub lexical: usize,
    pub character: usize,
}

impl Verdict {
    /// Unanimity was measured at 90.3% correct against 50% when split. The
    /// threshold is not a tuned parameter, it is the only place the curve has
    /// a step in it.
    pub fn confident(&self) -> bool {
        self.agreement == 3
    }
}
